### PR TITLE
[FLINK-31664][Table SQL/API] Add ARRAY_INTERSECT supported in SQL & Table API

### PR DESCRIPTION
**Pull request by zzzzzzzs**
## What is the purpose of the change

Add `array_intersect(array1, array2)` SQL function.

- `array1`: An ARRAY of any type with comparable elements.
- `array2`: n ARRAY of elements sharing a least common type with the elements of `array1`.

Returns an ARRAY of matching type to array1 with no duplicates and elements contained in both array1 and array2.

Examples:
```sql
SELECT array_intersect(array(1, 2, 3), array(1, 3, 3, 5));
[1,3]
```
See also
https://spark.apache.org/docs/latest/api/sql/index.html#array_intersect

## Brief change log

Add ARRAY_INTERSECT supported in SQL & Table API

## Verifying this change

This change is already covered by existing tests, such as CollectionFunctionsITCase#arrayIntersectTestCases.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayIntersectFunction.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+                ((CollectionDataType) context.getCallContext().getArgumentDataTypes().get(0))
+                        .getElementDataType();
+        elementGetter = ArrayData.createElementGetter(elementDataType.getLogicalType());
+    }
+
+    public @Nullable ArrayData eval(ArrayData haystack, ArrayData needle) {
+        int len1 = haystack.size();
+        int len2 = needle.size();
+        Object[] rs = new Object[Math.min(len1, len2)];
+        Set<Object> hs = new HashSet<>();
```
Comment (snuyanzin) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayIntersectFunction.java:52]: We shouldn't depend on java objects equality, 
instead it's better to use a generated comparator,
for more details please have a look e.g. at `array_contains`

File: docs/data/sql_functions.yml [hunk 2]
```diff
@@ -643,6 +643,9 @@ collection:
   - sql: ARRAY_REVERSE(haystack)
     table: haystack.arrayReverse()
     description: Returns an array in reverse order. If the array itself is null, the function will return null.
+  - sql: ARRAY_INTERSECT(haystack, needle)
+    table: haystack.arrayIntersect(needle)
+    description: Returns an array that contains the common elements of both input arrays.
```
Comment (snuyanzin) [inline docs/data/sql_functions.yml:648]: it doesn't reflect the fact mentioned in PR's description about only one element in common.
E.g. `array1 = array[1, 1, 2, 2, 3, 3]` and `array2 = array[2, 3, 2, 3]`
what will the function return?

also probably need to change name of args (`needle` and `haystack`)

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java [hunk 3]
```diff
@@ -1396,6 +1397,17 @@ public OutType arrayReverse() {
         return toApiSpecificExpression(unresolvedCall(ARRAY_REVERSE, toExpr()));
     }
 
+    /**
+     * Returns an array that contains the common elements of both input arrays.
```
Comment (snuyanzin) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java:1401]: same about unique elements in result

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java [hunk 4]
```diff
@@ -1396,6 +1397,17 @@ public OutType arrayReverse() {
...
+     *
+     * <p>Null values are not considered equal. If any of the input arrays is null, the function
```
Comment (snuyanzin) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java:1403]: why nulls are not considered to be equal?
what's other systems behavior?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java [hunk 5]
```diff
@@ -261,6 +261,21 @@ ANY, and(logical(LogicalTypeRoot.BOOLEAN), LITERAL)
                             "org.apache.flink.table.runtime.functions.scalar.ArrayReverseFunction")
                     .build();
 
+    public static final BuiltInFunctionDefinition ARRAY_INTERSECT =
+            BuiltInFunctionDefinition.newBuilder()
+                    .name("ARRAY_INTERSECT")
+                    .kind(SCALAR)
+                    .inputTypeStrategy(
+                            sequence(
+                                    Arrays.asList("haystack", "needle"),
```
Comment (snuyanzin) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:270]: looks like name of args are not suitable for this function

**Comment by snuyanzin** [comment]
sorry for being late  here 
probably you need to rebase the PR.

Also it seems the test is failed because you also need to update python related part for new function. Have a look at other similar implementations e.g. `ARRAY_CONTAINS`

**Review by snuyanzin — CHANGES_REQUESTED** [review]
some more findings:
these queries are failing with NPE
```sql
select array_intersect(array[1, 2, 3], cast(null as array<int>));
select array_intersect(cast(null as array<int>), array[1]);
select array_intersect(cast(null as array<int>), cast(null as array<int>));
```

This query returns strange result:
it returns `array[NULL]`
however arrays are of different types... I would first have a look what other vendors do in such situation
```sql
select array_intersect(array[cast(null as int)], array[cast(null as char)]);
```

Also it is still a question: why should we allow to apply this function to array of different types? It is ok if one of them could be implicitly casted to another however i don't see a big sense if it couldn't.

another issue as a result of allowance of different array types
a query 
```sql
select array_intersect(array[array[1, 2], array[1, 2]], array['a']);
```
fails with `java.lang.IndexOutOfBoundsException`
and a query
```sql
select array_intersect(array[array[1, 2], array[1, 2]], array[2, 3]);
```
fails with `java.lang.ArrayIndexOutOfBoundsException: 1`

**Comment by liuyongvs** [comment]
hi @zzzzzzzs @snuyanzin  do you still does it? if not , i will go on

**Comment by zzzzzzzs** [comment]
> hi @zzzzzzzs @snuyanzin do you still does it? if not , i will go on

I will continue in the near future

**Comment by MartijnVisser** [comment]
@zzzzzzzs Are you still working on this PR, or should someone take this over?

**Comment by dawidwys** [comment]
Since we have not heard in a while from the contributor and we have more advanced solutions, I'll close this PR.

* https://github.com/apache/flink/pull/23171
* https://github.com/apache/flink/pull/22629

**Pull request closed without merging**